### PR TITLE
Use log4js as logging library to get structured log metadata

### PR DESCRIPTION
**Pull request by lulf**
I've set the log level to info for now and not made use of any component names, but I can do that in a
second iteration if you think the choice of library is ok.

### CODE REVIEW

**Comment by grs** [comment]
I would rather use 'debug'  I think, since our needs are simple and it has no other dependencies (and is already used by rhea). We can pull the log interface into a simple module in our code and that would make it easy to change in future if needed (or to make pluggable).

Using debug will give us timestamps and the ability to selectively turn different things on based on env vars.

**Comment by lulf** [comment]
@grs Added some new changes.

**Comment by grs** [comment]
While I recognise that the usual level+category gives great flexibility, thinking aloud I wonder whether we need that. Might a very simple 3 level system do for each component? E.g. normal, debug and trace?

The normal logs messages would always be on and would include errors, warnings etc and probably also a record of each request and its outcome. The debug level would often be on in development and would provide a bit more detail on the various requests. The trace level would include everything.

My gut instinct is that this would be a more useful operational interface as there is no need to try to understand which categories might be relevant to the problem at hand. YOu would just have the ability to increase the verbosity to debug or trace.

The counter argument would be that the logs were too unwieldy at the extra levels of verbosity without further filtering. Perhaps though it would be worth trying this simpler form and get some practical experience on whether this is the case (in which case we would introduce some filtering, hopefully with more understanding of the categorisation that is practically useful).

Wdyt?

**Comment by lulf** [comment]
Just saw your comment before I pushed. In the latest commit I've made a simple logging lib that provides error, warn, info and debug levels. 

error: critical messages preventing the component to be used
warn: error messages that can be recovered from
info: messages that is not spammy and generally useful when looking at the logs
debug: development messages

I can add a trace level as well, which would make it similar to the levels used by java logging frameworks.

**Comment by grs** [comment]
I think this is good! A couple of comments:

At present, the DEBUG env var controls the category/component, and setting that in effect overrides the log level setting. I.e. if I set DEBUG=subserv:artemis, then I won't get any log output from other categories regardless of level  (and if I don't set DEBUG at all, I won't see any output). This seems somehow wrong to me.  Ideally I think it should not be possible to turn error and warning messages off at all. I think the log level should be the primary interface.

I don't think the current module names make the best logical categorisation, either for filtering on via DEBUG or for any post processing/analysis. (Also fair to say the log messages themselves may not be quite right of course).

In most cases, I think an extra level of verbosity would be more helpful than categories. I.e. by default you get info, warn and error. You can up the verbosity to get debug or up it again to turn absolutely everything on.

So perhaps we could (a) check DEBUG in the code and set it if not set, (b) make the subcategory optional, with the main category taken from process.argv[1]. That way we can still get more refined categorisation if we need it and even control log output based on it, but for the most part need only deal with one control - log level / verbosity - and the default would be as expected.

Wdyt? Sorry to be really picky here! We can also modify it as we go if you prefer, I'm just saying the sorts of changes I would have in mind.

**Comment by lulf** [comment]
I think your suggestion is good,  but just so I understand:

1. Move the setting of DEBUG from the dockerfiles to the process and only set it if not set already, to subserv:*
2. If DEBUG is not set externally and if argv[1] is passed, set it to subserv:argv[1]

The reason I'm not setting it to * is that rhea will give a lot of debug output as well, which can't be controlled by the log level.

**Comment by grs** [comment]
For node, argv[1] is the script name so subserv.js in this instance. MY thought was to use that by default so that the log module can eventually be used by other deamons/executables if needed and by default will use the name of the process as the base category (with any further subcategories being optional).

For the other part, yes we would set the DEBUG env var if not already set to: argv[1] + '.*'.

That way the DEBUG var is mostly 'hidden' or not required, but if for some reason you want to set it you can still do so, e.g. to enable some extra rhea info or whatever.

**Comment by lulf** [comment]
@grs Updates with requested changes.

File: Dockerfile [hunk 1]
```diff
@@ -8,7 +8,7 @@ RUN yum -y update  \
 
 RUN mkdir -p /opt/app-root/src/
 RUN cd /opt/app-root/src/
-RUN ["/bin/bash", "-c", "npm install debug && npm install bluebird && npm install rhea"]
+RUN ["/bin/bash", "-c", "npm install debug && npm install bluebird && npm install debug && npm install rhea"]
```
Comment (grs) [inline Dockerfile]: we don't need the extra debug (though I don't think it will in fact have any effect anyway)

**Pull request merged**